### DISCUSSION TITLE
Module "textual.app" does not explicitly export attribute "CSSPathType"

### DISCUSSION THREAD

**Discussion by tconbeer** [Textualize/textual#2817] [Q&A] [answered]
There was a type checking regression in v0.28.0 that was introduced by [this commit](https://github.com/Textualize/textual/commit/b98e004b5445396a32cf073bb679eb0e0781ae03), which imported `CSSPathType` in `app.py` from `_path.py`, even though that was defined as a constant lower in the file. Since the import isn't exported, it removes the public interface for CSSPathType, which is needed to subclass `App` and replace `App.__init__()`:

mypy doesn't like this. This will raise a type checking error in 0.28 (but not 0.27):

```
from textual.app import App, CSSPathType # error: Module "textual.app" does not explicitly export attribute "CSSPathType"  [attr-defined]

class MyApp(App, inherit_bindings=False):
    def __init__(
        self,
        foo: str,
        driver_class: Union[Type[Driver], None] = None,
        css_path: Union[CSSPathType, None] = None,
        watch_css: bool = False,
    ):
        super().__init__(driver_class, css_path, watch_css)
        self.foo = foo
```

**Comment by davep** [comment, marked as answer]
`CSSPathType` [now lives in `textual.types`](https://textual.textualize.io/api/types/#textual.types.CSSPathType). The likes of `mypy` and `pyright` seem happy with this kind of code:

```python
from __future__ import annotations

from textual.app import App, ComposeResult
from textual.widgets import Label
from textual.types import CSSPathType

class MyApp(App[None]):

    def __init__(self, css_path: CSSPathType | None ) -> None:
        super().__init__(css_path=css_path)

    def compose(self) -> ComposeResult:
        yield Label("Hello, World!")

if __name__ == "__main__":
    MyApp("foo.css").run()
```

**Reply by tconbeer** [reply]
Thanks. Couldn't find it with a code search

**Reply by davep** [reply]
For what it's worth, I only found the correct location by scanning through the PR you linked. ;-)

![Screenshot 2023-06-21 at 13 47 32](https://github.com/Textualize/textual/assets/28237/649a85ee-71ff-4ea8-90a5-d52d889651d1)

Looks like making this change clear was something left out of the CHANGELOG, so easily missed.

**Reply by willmcgugan** [reply]
Probably because it may have seemed like an internal detail at the time. You're right though. We should document these in the changelog.
